src/System.Device.Gpio/System/Device/I2c/UnixI2cBus.cs

```diff
@@ -0,0 +1,218 @@
+                {
+                    throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not open I2C device file '{deviceFileName}'.");
+                }
+
+                I2cFunctionalityFlags functionalityFlags;
+                int result = Interop.ioctl(busFileDescriptor, (uint)I2cSettings.I2C_FUNCS, new IntPtr(&functionalityFlags));
+                if (result < 0)
+                {
+                    functionalityFlags = 0;
+                }
+
+                if ((functionalityFlags & I2cFunctionalityFlags.I2C_FUNC_I2C) != 0)
+                {
+                    return new UnixI2cBus(busFileDescriptor, busId);
+                }
+                else
+                {
+                    return new UnixI2cFileTransferBus(busFileDescriptor, busId);
+                }
+            }
+        }
+
+        public UnixI2cBus(int busFileDescriptor, int busId)
+        {
+            BusId = busId;
+            BusFileDescriptor = busFileDescriptor;
+        }
+
+        public override I2cDevice CreateDevice(int deviceAddress)
+        {
+            if (!HasValidFileDescriptor())
+            {
+                throw new ObjectDisposedException(nameof(UnixI2cBus));
+            }
+
+            _usedAddresses ??= new HashSet<int>();
+            if (!_usedAddresses.Add(deviceAddress))
+            {
+                throw new ArgumentException($"Device with address 0x{deviceAddress,0X2} is already open.", nameof(deviceAddress));
+            }
+
+            return CreateDeviceNoCheck(deviceAddress);
+        }
+
+        internal I2cDevice CreateDeviceNoCheck(int deviceAddress)
+        {
+            return new UnixI2cDevice(this, deviceAddress);
+        }
+
+        public override void RemoveDevice(int deviceAddress)
+        {
+            if (!RemoveDeviceNoCheck(deviceAddress))
+            {
+                throw new ArgumentException($"Device with address 0x{deviceAddress,0X2} was not open.", nameof(deviceAddress));
+            }
+        }
+
+        internal bool RemoveDeviceNoCheck(int deviceAddress)
+        {
+            return _usedAddresses?.Remove(deviceAddress) ?? false;
+        }
+
+        public override unsafe void Read(int deviceAddress, Span<byte> buffer)
+        {
+            if (deviceAddress < 0 || deviceAddress > ushort.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(deviceAddress));
+            }
+
+            if (buffer.Length == 0)
+            {
+                throw new ArgumentException($"{nameof(buffer)} cannot be empty.", nameof(buffer));
+            }
+
+            if (buffer.Length > ushort.MaxValue)
+            {
+                throw new ArgumentException($"{nameof(buffer)} length is too long.", nameof(buffer));
+            }
+
+            fixed (byte* readBufferPointer = buffer)
+            {
+                WriteReadCore((ushort)deviceAddress, null, readBufferPointer, 0, (ushort)buffer.Length);
```

Comment:
>WriteReadCore [](start = 16, length = 13)

This is another problem of having the read/write/readwrite methods here. With this we now have two almost identical implementations of all of these methods, one here and one on UnixI2cDevice so code duplication will just be prone to issues when fixing bugs as we may fix it in one place but not the other. I would strongly recommend to just remove all these operations from here and leave them on I2cDevice, or worst case scenario keep a list of I2cDevices opened on the bus class and simply calling i2cDevice.Read from here.

Reply:
UnixI2cDevice doesn't duplicate this code

Reply:
oh nvm, now I see that you also changed the I2cDevice to now use the bus to perform the operations, so at least we won't have code duplication. I still have the concern of having to classes doing almost the same in terms of API surface.

Reply:
I2cDevice is a convenience and mostly recommended one. I2cBus should only be used when you get very specialized APIs. I.e.:
- SenseHat which has 6 devices and chaining it with Ft4222 is a pure pain without I2cBus
- Ft4222 only has a concept of a bus which means it needs an object to represent it, while you could create I2cBus just for Ft4222 it just makes it easier to use in context like SenseHat
- https://github.com/dotnet/iot/issues/214 uses different address depending on memory size
- I2C scanning

for case where you use a single address you should only use I2cDevice. I2cBus overload could be provided for convenience so that address doesn't have to be passed

Reply:
Add to this: 1024 addressing support, SMBus specific support (some will still be missing but getting there), broadcast to the bus, hardware I2C programming, more information here: https://www.i2c-bus.org/addressing/